### PR TITLE
feat(compass-aggregations): export aggregations COMPASS-5543

### PR DESCRIPTION
**Pull request by leorossi**
## Description
Shows "Export" button in Aggregations Toolbar, to export aggregation results in JSON/CSV file.

To enable that run compass with command
```
env COMPASS_SHOW_NEW_AGGREGATION_TOOLBAR=true COMPASS_ENABLE_AGGREGATION_EXPORT=true npm start
```
Screenshot_2022-04-07_09-38-27

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context

- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-import-export/src/components/export-modal/export-modal.jsx [hunk 1]
```diff
@@ -280,6 +281,7 @@ class ExportModal extends PureComponent {
         exportedDocsCount={this.props.exportedDocsCount}
         selectExportFileType={this.props.selectExportFileType}
         selectExportFileName={this.props.selectExportFileName}
+        isAggregation={this.props.aggregation}
```
Comment (mabaasit) [inline packages/compass-import-export/src/components/export-modal/export-modal.jsx:284]: Passed here as an object but the component requires it as a boolean.

File: packages/compass-import-export/src/components/export-modal/export-modal.jsx [hunk 2]
```diff
@@ -84,6 +84,7 @@ class ExportModal extends PureComponent {
     toggleFullCollection: PropTypes.func.isRequired,
     selectExportFileType: PropTypes.func.isRequired,
     selectExportFileName: PropTypes.func.isRequired,
+    aggregation: PropTypes.object,
```
Comment (mabaasit) [inline packages/compass-import-export/src/components/export-modal/export-modal.jsx:87]: in state we set it as `false` by default. maybe we can make it optional there?

Comment (leorossi) [reply packages/compass-import-export/src/components/export-modal/export-modal.jsx:87]: Changing it to default `{}` for better consistency

File: packages/compass-import-export/src/components/progress-bar/progress-bar.jsx [hunk 3]
```diff
@@ -102,17 +107,23 @@ class ProgressBar extends PureComponent {
...
+    } = this.props;
 
     const formattedWritten = formatNumber(docsWritten);
     const formattedTotal = formatNumber(docsTotal);
 
     return (
       <p
         className={style('status-stats')}
-        title={progressTitle(formattedWritten, formattedTotal)}
+        title={progressTitle(formattedWritten, formattedTotal, isAggregation)}
```
Comment (mabaasit) [inline packages/compass-import-export/src/components/progress-bar/progress-bar.jsx:124]: Is this third param (`progressTitle`) extra here? The signature of `progressTitle` did not change anywhere

File: packages/compass-import-export/src/components/export-modal/export-modal.jsx (hunk 2 above)
Comment (mabaasit) [reply packages/compass-import-export/src/components/export-modal/export-modal.jsx:87]: in that case `!!this.props.aggregation` won't return false if its empty

File: packages/compass-import-export/src/modules/export.js [hunk 4]
```diff
@@ -426,70 +439,84 @@ export const startExport = () => {
...
-        .some(
-          (_part, index, parts) =>
-            projection[parts.slice(0, index + 1).join('.')]
+    const dest = fs.createWriteStream(exportData.fileName);
+    let columns, source, numDocsToExport;
+    let numDocsActuallyExported = 0;
+    if (!isEmpty(exportData.aggregation)) {
+      // Start Aggregation export
+      const { stages, options } = exportData.aggregation;
+      // // Get Document Fields
```
Comment (mabaasit) [inline packages/compass-import-export/src/modules/export.js:448]: leftover?

**Review by mabaasit — APPROVED** [review]

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx [hunk 5]
```diff
@@ -32,16 +32,31 @@ type PipelineActionsProps = {
...
 };
 
 export const PipelineActions: React.FunctionComponent<PipelineActionsProps> = ({
   isOptionsVisible,
   onRunAggregation,
   onToggleOptions,
+  onExportAggregationResults
 }) => {
   const optionsIcon = isOptionsVisible ? 'CaretDown' : 'CaretRight';
+  const showExportButton = process.env.COMPASS_ENABLE_AGGREGATION_EXPORT === 'true'
```
Comment (Anemy) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx:45]: Since `compass-aggregations` is embedded in cloud/browser, how do you feel about making these `process.env` checks undefined safe:
```suggestion
  const showExportButton = process?.env?.COMPASS_ENABLE_AGGREGATION_EXPORT === 'true'
```

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 6]
```diff
@@ -224,4 +225,32 @@ const fetchAggregationData = (page: number): ThunkAction<
...
+      namespace,
+      maxTimeMS,
+      collation,
+    } = getState();
+    
+    const stages = pipeline
+      .map(generateStage)
+      .filter((stage) => Object.keys(stage).length > 0);
+    const options: AggregateOptions = {
+      maxTimeMS: maxTimeMS || DEFAULT_MAX_TIME_MS,
```
Comment (Anemy) [inline packages/compass-aggregations/src/modules/aggregation.ts:246]: Should we allow users enter a value of `0` for `maxTimeMS` and use it? I'm thinking if users want the[ unbounded default behavior](https://www.mongodb.com/docs/manual/reference/command/aggregate/) they might put in a value of `0` which would be overridden here.

File: packages/compass-import-export/src/modules/export.js [hunk 7]
```diff
@@ -426,70 +439,73 @@ export const startExport = () => {
...
-    const columns = Object.keys(exportData.fields).filter((field) =>
-      field
-        .split('.')
-        .some(
-          (_part, index, parts) =>
-            projection[parts.slice(0, index + 1).join('.')]
+    const dest = fs.createWriteStream(exportData.fileName);
+    let columns, source, numDocsToExport;
+    let numDocsActuallyExported = 0;
+    if (!isEmpty(exportData.aggregation)) {
```
Comment (Anemy) [inline packages/compass-import-export/src/modules/export.js:445]: How do you feel about breaking the code in these condition blocks into functions? I'm thinking it might improve readability.

Maybe something like 
```ts
const {
  columns,
  source,
  numDocsToExport
} = getAggregationExportSource(dataService, ns, exportData);
```

**Review by Anemy — COMMENTED** [review]
This is really cool, I like how we're sharing the export flow. Left a couple questions/comments on a possible edge case and maybe a way we could improve readability.

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx (hunk 5 above)
Comment (Anemy) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx:45]: Thinking about this more I wonder if there's a way we could be testing the `compass-aggregations` (and maybe some other packages) on more real browser test environments to surface things like this more. Maybe worth a ticket down the line?

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.spec.tsx [hunk 8]
```diff
@@ -33,6 +37,26 @@ describe('PipelineActions', function () {
...
+        />
+      );
+      const button = screen.getByTestId('pipeline-toolbar-export-aggregation-button');
+      expect(button).to.exist;
+
+      userEvent.click(button);
+
+      expect(onExportAggregationResultsSpy.calledOnce).to.be.true;
+      expect(onExportAggregationResultsSpy.firstCall.args).to.be.empty;
+      delete process.env.COMPASS_ENABLE_AGGREGATION_EXPORT
```
Comment (mcasimir) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.spec.tsx:57]: This is a super nit since this is just a flag specific to this feature, but it would be more accurate to do this inside an `afterEach` and restore the value present before the test run, rather than delete the variable.

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx (hunk 5 above)
Comment (gribnoysup) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx:45]: It's worth having the optional chaining here, but this will be processed by webpack (twice) and by default webpack will provide `process` for the browser environment (or in some cases will remove these expressions based on the configuration, although I'm not sure if optional chaining is something they handle in this case)

File: packages/compass-import-export/src/components/export-modal/export-modal.jsx (hunk 2 above)
Comment (gribnoysup) [inline packages/compass-import-export/src/components/export-modal/export-modal.jsx:87]: When mapping state to props you always want to make sure that you are passing what is required by the view, and not just passing state directly to the components. If this view only needs to know if aggregation exists or not, the logic for deriving this value should probably leave outside the component

File: packages/compass-import-export/src/modules/export.js [hunk 9]
```diff
@@ -126,12 +128,18 @@ const reducer = (state = INITIAL_STATE, action) => {
   }
 
   if (action.type === ON_MODAL_OPEN) {
-    return {
+    const newState = {
       ...INITIAL_STATE,
       count: action.count,
       query: action.query,
+      aggregation: action.aggregation || {},
```
Comment (gribnoysup) [inline packages/compass-import-export/src/modules/export.js:135]: Empty object is a pretty inconvenient value to represent "empty" that is hard to check in the other code, I'd suggest to have it set to `null` instead (similar for the query). Also as you only can have one or another, maybe it's worth having some sort of "type" value in the state

File: packages/compass-import-export/.gitignore [hunk 10]
```diff
@@ -0,0 +1 @@
+lib
```
Comment (gribnoysup) [inline packages/compass-import-export/.gitignore:1]: I don't think we need this anymore after clean-up that @Anemy did

File: packages/compass-import-export/src/components/export-modal/export-modal.jsx [hunk 11]
```diff
@@ -387,7 +390,7 @@ class ExportModal extends PureComponent {
           )}
         </Modal.Body>
         <Modal.Footer>
-          {this.renderBackButton()}
+          {this.props.isAggregation ? '' : this.renderBackButton()}
```
Comment (gribnoysup) [inline packages/compass-import-export/src/components/export-modal/export-modal.jsx:393]: Total nit, but this is a more common way to do this in React

```suggestion
          {!this.props.isAggregation && this.renderBackButton()}
```

File: packages/compass-import-export/src/modules/export.js [hunk 12]
```diff
@@ -277,11 +284,12 @@ export const queryChanged = (query) => ({
  * @param {Number} document count given current query.
  * @param {Object} current query.
  */
-export const onModalOpen = ({ namespace, count, query }) => ({
+export const onModalOpen = ({ namespace, count, query, aggregation }) => ({
   type: ON_MODAL_OPEN,
   namespace,
   count,
-  query,
+  query: query || {},
```
Comment (gribnoysup) [inline packages/compass-import-export/src/modules/export.js:291]: Same as for aggregation, empty object is not a great default / empty value due to ambiguity

File: packages/compass-import-export/src/modules/export.js [hunk 13]
```diff
@@ -362,19 +370,23 @@ const fetchDocumentCount = async (dataService, ns, query) => {
...
+}) => {
   return async (dispatch, getState) => {
     track('Export Opened');
     const {
       dataService: { dataService },
     } = getState();
-
     try {
       const count =
         maybeCount ?? (await fetchDocumentCount(dataService, namespace, query));
```
Comment (gribnoysup) [inline packages/compass-import-export/src/modules/export.js:386]: Not great that we are relying there on the fact that we pass zero as a aggregation count to avoid this fetch to happen. This creates an interface for the event that is hard to reason about. If we don't want to fetch document count for aggregation, we should be explicit about it

File: packages/compass-import-export/src/modules/export.js [hunk 14]
```diff
@@ -126,12 +128,18 @@ const reducer = (state = INITIAL_STATE, action) => {
...
       count: action.count,
-      query: action.query,
+      query: action.query || null,
+      aggregation: action.aggregation || null,
       isOpen: true,
     };
+
+    if (action.aggregation) {
+      newState.exportStep = EXPORT_STEP.FILETYPE;
+    }
```
Comment (gribnoysup) [inline packages/compass-import-export/src/modules/export.js:141]: I think the plan was to still have one screen that will show the aggregation that will be exported and this doesn't look like it's happening right now: we skip right to file type selection. Was there a decision to change the plan or are we planning to do this in a separate PR?

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx [hunk 15]
```diff
@@ -32,16 +32,32 @@ type PipelineActionsProps = {
...
+        <Button
+          data-testid="pipeline-toolbar-export-aggregation-button"
+          variant="default"
+          size="small"
+          onClick={() => {
+            onExportAggregationResults();
+          }}
+        >
+          Export
+        </Button>
```
[the remark this answers is not in the document]
Comment (mabaasit) [reply packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/pipeline-actions.tsx:59]: Created a follow up: [COMPASS-5734](https://jira.mongodb.org/browse/COMPASS-5734)

File: packages/compass-import-export/src/modules/export.js (hunk 14 above)
Comment (mabaasit) [reply packages/compass-import-export/src/modules/export.js:141]: Thanks for pointing that out. I created a follow up: [COMPASS-5735](https://jira.mongodb.org/browse/COMPASS-5735)

File: packages/compass-import-export/src/modules/export.js [hunk 16]
```diff
@@ -426,70 +447,23 @@ export const startExport = () => {
...
-      length: numDocsToExport,
-      time: 250 /* ms */,
-    });
-
-    progress.on('progress', function (info) {
-      dispatch(onProgress(info.percentage, info.transferred));
-    });
-
-    log.info(mongoLogId(1001000084), 'Export', 'Start writing to file', {
+    const dest = fs.createWriteStream(exportData.fileName);
```
Comment (Anemy) [inline packages/compass-import-export/src/modules/export.js:450]: We used to create this write stream and the export source in the `try` - is there a reason we want to move it out? Is there any way it could fail that we want to catch? Maybe nothing to change.

**Review by Anemy — APPROVED** [review]
lgtm! Thanks for picking this up. I left one comment on possible error catching

**Review by gribnoysup — APPROVED** [review]
With the follow-up tickets created and as this is behind a feature flag this looks good to be merged. Let's just make sure that CI is green before doing that
